dm/worker/meta.go

```diff
@@ -74,95 +77,240 @@ func (m *Meta) adjust() error {
 	return nil
 }
 
-// FileMetaDB stores meta information in disk
-type FileMetaDB struct {
-	lock sync.Mutex // we need to ensure only a thread can access to `metaDB` at a time
-	meta *Meta
-	path string
-}
+// Metadata stores metadata and log of task
+// it also provides logger feature
+// * append log
+// * forward to specified log location
+type Metadata struct {
+	sync.RWMutex // we need to ensure only a thread can access to `metaDB` at a time
+	wg           sync.WaitGroup
+	ctx          context.Context
+	cancel       context.CancelFunc
 
-// NewFileMetaDB returns a meta file db
-func NewFileMetaDB(dir string) (*FileMetaDB, error) {
-	metaDB := &FileMetaDB{
-		path: path.Join(dir, "meta"),
-		meta: &Meta{
-			SubTasks: make(map[string]*config.SubTaskConfig),
-		},
-	}
+	// cache
+	tasks map[string]*pb.TaskMeta
+	logs  []*pb.TaskLog
+
+	// task operation log
+	log *Logger
+
+	dir string
+	db  *leveldb.DB
+}
 
-	if err := os.MkdirAll(dir, 0700); err != nil {
-		return nil, errors.Annotatef(err, "create meta directory %s", dir)
+// NewMetadata returns a metadata object
+func NewMetadata(dir string, db *leveldb.DB) (*Metadata, error) {
+	meta := &Metadata{
+		dir: dir,
+		db:  db,
+		log: new(Logger),
 	}
 
-	fd, err := os.Open(metaDB.path)
-	if os.IsNotExist(err) {
-		log.Warnf("failed to open meta file %s, going to create a new one: %v", metaDB.path, err)
-		return metaDB, nil
-	} else if err != nil {
+	// restore from old metadata
+	oldPath := path.Join(dir, "meta")
+	fd, err := os.Open(oldPath)
+	// old metadata file exists, recover metadata from it
+	if err == nil {
+		fd.Close()
```

Comment:
can we move this `Open`/`Close` into `recoverMetaFromOldFashion`? and do all of these (try/do) in `recoverMetaFromOldFashion`?

Reply:
Sure